lib/msal-node/src/client/ManagedIdentitySources/CloudShell.ts

```diff
@@ -3,17 +3,107 @@
  * Licensed under the MIT License.
  */
 
-import { ManagedIdentityId } from "../../config/ManagedIdentityId";
-import { ManagedIdentityRequestParameters } from "../../config/ManagedIdentityRequestParameters";
+import {
+    CacheManager,
+    INetworkModule,
+    Logger,
+    UrlString,
+} from "@azure/msal-common";
 import { BaseManagedIdentitySource } from "./BaseManagedIdentitySource";
+import {
+    MANAGED_IDENTITY_MSI_ENDPOINT,
+    HttpMethod,
+} from "../../utils/Constants";
+import { CryptoProvider } from "../../crypto/CryptoProvider";
+import { ManagedIdentityRequestParameters } from "../../config/ManagedIdentityRequestParameters";
+import {
+    ManagedIdentityErrorCodes,
+    createManagedIdentityError,
+} from "../../error/ManagedIdentityError";
 
 export class CloudShell extends BaseManagedIdentitySource {
-    public createRequest(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _resource: string,
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _managedIdentityId: ManagedIdentityId
-    ): ManagedIdentityRequestParameters {
-        throw new Error("Method not implemented.");
+    private endpoint: string;
+
+    constructor(
+        logger: Logger,
+        cacheManager: CacheManager,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider,
+        endpoint: string
+    ) {
+        super(logger, cacheManager, networkClient, cryptoProvider);
+
+        this.endpoint = endpoint;
+    }
+
+    public static tryCreate(
+        logger: Logger,
+        cacheManager: CacheManager,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider
+    ): CloudShell | null {
+        const [areEnvironmentVariablesValidated, endpoint]: [
+            boolean,
+            string | undefined
+        ] = validateEnvironmentVariables(
+            process.env[MANAGED_IDENTITY_MSI_ENDPOINT] || undefined,
+            logger
+        );
+
+        return areEnvironmentVariablesValidated
+            ? new CloudShell(
+                  logger,
+                  cacheManager,
+                  networkClient,
+                  cryptoProvider,
+                  endpoint as string
+              )
+            : null;
+    }
+
+    public createRequest(): ManagedIdentityRequestParameters {
+        const request: ManagedIdentityRequestParameters =
+            new ManagedIdentityRequestParameters(
+                HttpMethod.POST,
+                this.endpoint
+            );
+
+        request.headers["ContentType"] = "application/x-www-form-urlencoded";
+        request.headers["Metadata"] = "true";
+        // bodyParameters calculated in BaseManagedIdentity.acquireTokenWithManagedIdentity
```

Comment:
Can you confirm that resource is added as a body parameter?

Reply:
No, it's a query parameter. Look at the IMDS code.

Reply:
Ok, I see it but in the Cloud Shell implementation in .NET it is in the body parameters
`request.BodyParameters.Add("resource", resource);`

Shouldn't it also be a body parameter here?

Reply:
As it is now, it eventually gets passed to the NodeJS HttpClient as 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://graph.microsoft.com'.

However, I just remembered that CloudShell sends post requests, and is the only Managed Identity source to make a post request instead of a get request. Therefore, I think you'll need to add the resource to the body in the if-statement in BaseManagedIdentitySource.ts.